ksqldb-rest-app/src/main/java/io/confluent/ksql/logging/query/QueryAnonymizingRewritePolicy.java

```diff
@@ -0,0 +1,91 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.logging.query;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.parser.ParsingException;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.QueryAnonymizer;
+import io.confluent.ksql.util.QueryGuid;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Stream;
+import org.apache.log4j.Logger;
+import org.apache.log4j.rewrite.RewritePolicy;
+import org.apache.log4j.spi.LoggingEvent;
+
+public final class QueryAnonymizingRewritePolicy implements RewritePolicy {
+  private static final QueryAnonymizer anonymizer = new QueryAnonymizer();
+  private final String namespace;
+  private final Boolean anonymizeQueries;
+
+  public QueryAnonymizingRewritePolicy(final KsqlConfig config) {
+    final String clusterNamespace =
+        config.getString(KsqlConfig.KSQL_CCLOUD_QUERYANONYMIZER_CLUSTER_NAMESPACE);
+    this.namespace =
+        clusterNamespace.isEmpty()
+            ? config.getString(KsqlConfig.KSQL_SERVICE_ID_CONFIG)
+            : clusterNamespace;
+    this.anonymizeQueries = config.getBoolean(KsqlConfig.KSQL_CCLOUD_QUERYANONYMIZER_ENABLED);
+  }
+
+  private QueryGuid buildGuids(final String query, final String anonymizedQuery) {
+    return new QueryGuid(namespace, query, anonymizedQuery);
+  }
+
+  @Override
+  public LoggingEvent rewrite(final LoggingEvent source) {
+    final ImmutableMap<String, String> oldMessage =
+        (ImmutableMap<String, String>) source.getMessage();
+    final String query = oldMessage.get("query");
+
+    final Map props = new HashMap(source.getProperties());
+
+    String anonymizedQuery;
+    try {
+      anonymizedQuery = anonymizer.anonymize(query);
+    } catch (ParsingException e) {
+      anonymizedQuery = "";
+    }
+
+    final QueryGuid queryGuids = buildGuids(query, anonymizedQuery);
+    final ImmutableMap<Object, Object> newMessage =
+        ImmutableMap.builder()
+            .put("query", anonymizeQueries ? anonymizedQuery : query)
+            .put("queryGUID", queryGuids.getQueryGuid())
+            .put("structuralGUID", queryGuids.getStructuralGuid())
+            .build();
+
+    final ImmutableMap<Object, Object> combined =
+        Stream.concat(oldMessage.entrySet().stream(), newMessage.entrySet().stream())
+            .collect(
+                toImmutableMap(Map.Entry::getKey, Map.Entry::getValue, (left, right) -> right));
+
+    return new LoggingEvent(
+        source.getFQNOfLoggerClass(),
+        source.getLogger() != null ? source.getLogger() : Logger.getLogger(source.getLoggerName()),
+        source.getTimeStamp(),
+        source.getLevel(),
+        combined,
```

Comment:
Might make sense to use an object that has a sane `toString` method in case someone wants to use this with an off-the-shelf appender.

Reply:
nvm, turns out `ImmutableMap.toString` _is_ sane!